src/java.base/share/classes/java/util/Formatter.java

```diff
@@ -2856,6 +2850,171 @@ static List<FormatString> parse(String s) {
+        final String s;
+        final int max;
+        char first;
+        int start;
+        int off;
+        char c;
+        int argSize;
+        int flagSize;
+        int widthSize;
+
+        FormatSpecifierParser(ArrayList<FormatString> al, char first, int start, String s, int max) {
+            this.al = al;
+
+            this.first = first;
+            this.c = first;
+            this.start = start;
+            this.off = start;
+
+            this.s = s;
+            this.max = max;
+        }
+
+        void reset(char first, int start) {
+            this.first = first;
+            this.c = first;
+            this.start = start;
+            this.off = start;
+
+            argSize = 0;
+            flagSize = 0;
+            widthSize = 0;
+        }
+
+        int parse() {
+            // %[argument_index$][flags][width][.precision][t]conversion
+            // %(\d+\$)?([-#+ 0,(\<]*)?(\d+)?(\.\d+)?([tT])?([a-zA-Z%])
+            int precisionSize = 0;
+
+           parseArgument();
+
+            if (widthSize == 0) {
+                if (flagSize == 0) {
+                    parseFlag();
+                }
+                parseWidth();
+            }
+
+            if (c == '.' && off + 1 < max) {
+                // (\.\d+)?
+                precisionSize = parsePrecision();
+                if (precisionSize == 0) {
+                    return 0;
+                }
+            }
+
+            // ([tT])?([a-zA-Z%])
+            char t = '\0', conversion = '\0';
+            if ((c == 't' || c == 'T') && off + 1 < max) {
+                char c1 = s.charAt(off + 1);
+                if (isConversion(c1)) {
+                    t = c;
+                    conversion = c1;
+                    off += 2;
+                }
+            }
+            if (conversion == '\0' && isConversion(c)) {
+                conversion = c;
+                ++off;
+            }
+
+            if (argSize + flagSize + widthSize + precisionSize + t + conversion != 0) {
+                if (al != null) {
+                    FormatSpecifier formatSpecifier
+                            = new FormatSpecifier(s, start, argSize, flagSize, widthSize, precisionSize, t, conversion);
+                    al.add(formatSpecifier);
+                }
+                return off - start;
+            }
+            return 0;
+        }
+
+        private void parseArgument() {
+            // (\d+\$)?
+            for (int size = 0; off < max; ++off, c = s.charAt(off), size++) {
+                if (!isDigit(c)) {
+                    if (size > 0) {
+                        if (c == '$') {
+                            ++off;
+                            argSize = size + 1;
+                            size = 0;
+                            if (off < max) {
+                                c = s.charAt(off);
+                            }
+                        } else {
+                            if (first == '0') {
```

Comment:
While it's clever to avoid re-parsing I think it muddies the control flow. It would be simpler if we always reset to `off = start; c = first` in this `else` block then unconditionally call `parseFlags(); parseWidth();` outside in `parse`. The few extra calls to `s.charAt(..)` this might add a little overhead on some tests, but the JIT might like the brevity and less branchy structure overall and on larger benchmarks.. Maybe worth experimenting with.

Reply:
Good idea. In addition, I also plan to simplify the writing of the for statement, such as:
```java
for (int size = 0; off < max; ++off, c = s.charAt(off), size++) {
==>
for (int size = 0; off < max; c = s.charAt(++off), size++) {
```